src/main/java/com/urbanairship/datacube/OffsetBasedAddressFormat.java

```diff
@@ -0,0 +1,231 @@
+                }
+                // add the bucket or the bucket id retrieved from the id service
+                reversedKeyElems.add(bucket);
+                // add the identifier for the bucket
+                reversedKeyElems.add(bucketTypeId);
+                // add add the indicator saying its not a wildcard
+                reversedKeyElems.add(NON_WILDCARD_FIELD);
+            }
+        }
+
+        // flip em
+        List<byte[]> keyElemsInOrder = Lists.reverse(reversedKeyElems);
+
+        int totalKeySize = 0;
+        for (byte[] keyElement : keyElemsInOrder) {
+            totalKeySize += keyElement.length;
+        }
+
+        ByteBuffer bb;
+        // Add a place holder for the hash byte if it's required
+        if (useAddressPrefixByteHash) {
+            bb = ByteBuffer.allocate(totalKeySize + 1);
+            byte hashByte = Util.hashByteArray(bb.array(), 1, totalKeySize + 1);
+            bb.put(hashByte);
+        } else {
+            bb = ByteBuffer.allocate(totalKeySize);
+        }
+
+        for (byte[] keyElement : keyElemsInOrder) {
+            bb.put(keyElement);
+        }
+
+        if (bb.remaining() != 0) {
+            throw new AssertionError("Key length calculation was somehow wrong, " +
+                    bb.remaining() + " bytes remaining");
+        }
+        return Optional.of(bb.array());
+    }
+
+    private byte[] getBucket(boolean readOnly, Dimension<?> dimension, BucketTypeAndBucket bucketAndCoord) throws IOException, InterruptedException {
+        byte[] bucket;
+
+        if (idService == null || !dimension.getDoIdSubstitution()) {
+            bucket = bucketAndCoord.bucket;
+        } else {
+            int dimensionNum = dimensions.indexOf(dimension);
+            if (readOnly) {
+                final Optional<byte[]> maybeId =
+                        idService.getId(dimensionNum, bucketAndCoord.bucket, dimension.getNumFieldBytes());
+
+                bucket = maybeId.orElse(null);
+
+                if (bucket == null) {
+                    return bucket;
+                }
+
+            } else {
+                bucket = idService.getOrCreateId(dimensionNum, bucketAndCoord.bucket, dimension.getNumFieldBytes());
+            }
+        }
+
+
+        if (bucket.length != dimension.getNumFieldBytes()) {
+            throw new IllegalArgumentException("Field length was wrong (after bucketing " +
+                    " and unique ID substitution). For dimension " + dimension +
+                    ", expected length " + dimension.getNumFieldBytes() + " but was " +
+                    bucket.length);
+        }
+        return bucket;
+    }
+
+    @Override
+    public Optional<Address> fromKey(byte[] bytes) throws IOException, InterruptedException {
+        int remainingBytes = bytes.length;
+
+        ByteBuffer bb = ByteBuffer.wrap(bytes);
+
+        if (useAddressPrefixByteHash) {
+            // then read the first int and throw it out.
+            bb.getInt();
```

Comment:
todo: might be cool to a) include the partition on the built address, b) compute the hashcode again on read to make sure we haven't got data corruption.